cpp/include/cudf/detail/iterator.cuh

```diff
@@ -43,6 +43,32 @@
 
 namespace cudf {
 namespace detail {
+/**
+ * @brief Convenience wrapper for creating a `thrust::transform_iterator` over a
+ * `thrust::counting_iterator`.
+ *
+ * Example:
+ * @code{.cpp}
+ * // Returns square of the value of the counting iterator
+ * auto iter = make_counting_transform_iterator(0, [](auto i){ return (i * i);});
+ * iter[0] == 0
+ * iter[1] == 1
+ * iter[2] == 4
+ * ...
+ * iter[n] == n * n
+ * @endcode
+ *
+ * @param start The starting value of the counting iterator
+ * @param f The unary function to apply to the counting iterator.
+ * This should be a host function and not a device function.
```

Comment:
Doesn't this depend on how the iterator is used? It certainly is being used with device functors in this PR.

Reply:
That is a good point. @jrhemstad is there a reason it says this?

Reply:
I think this was specific to when it was solely a test utility where the iterator was only ever invoked in host code.